src/org/intellij/erlang/completion/ErlangAppCompletionContributor.java

```diff
@@ -0,0 +1,98 @@
+    PsiFile file = context.getFile();
+    if (file.getFileType() != ErlangFileType.APP) return;
+    int startOffset = context.getStartOffset();
+    PsiElement elementAt = file.findElementAt(startOffset);
+    PsiElement parent = elementAt != null ? elementAt.getParent() : null;
+    if (parent instanceof ErlangTupleExpression) {
+      context.setDummyIdentifier("a");
+    }
+  }
+
+  public ErlangAppCompletionContributor() {
+    extend(CompletionType.BASIC, psiElement().inFile(instanceOf(ErlangFile.class))
+                                             .inVirtualFile(virtualFile().ofType(ErlangFileType.APP))
+                                             .with(parentsOfType(ErlangFile.class)),
+           new CompletionProvider<CompletionParameters>() {
+             @Override
+             protected void addCompletions(@NotNull CompletionParameters parameters,
+                                           ProcessingContext context,
+                                           @NotNull CompletionResultSet result) {
+               result.addElement(PrioritizedLookupElement.withPriority(LookupElementBuilder.create("application"),
+                                                                       ErlangCompletionContributor.KEYWORD_PRIORITY));
+             }
+           });
+    extend(CompletionType.BASIC, psiElement().inFile(instanceOf(ErlangFile.class))
+                                             .inVirtualFile(virtualFile().ofType(ErlangFileType.APP))
+                                             .with(parentsOfType(ErlangListExpression.class)),
+           new CompletionProvider<CompletionParameters>() {
+             @Override
+             protected void addCompletions(@NotNull CompletionParameters parameters,
+                                           ProcessingContext context,
+                                           @NotNull CompletionResultSet result) {
+               for (String keyword : KEYWORDS) {
+                 result.addElement(PrioritizedLookupElement.withPriority(LookupElementBuilder.create(keyword),
+                                                                         ErlangCompletionContributor.KEYWORD_PRIORITY));
+               }
+             }
+           });
+  }
+
+  @NotNull
+  private static PatternCondition<PsiElement> parentsOfType(@NotNull final Class<? extends PsiElement> oldestParent) {
+    return new PatternCondition<PsiElement>("withCorrectParents") {
+      @Override
+      public boolean accepts(@NotNull PsiElement element, ProcessingContext context) {
+        ErlangQAtom atom = PsiTreeUtil.getParentOfType(element, ErlangQAtom.class);
+        ErlangConfigExpression config = ObjectUtils.tryCast(atom != null ? atom.getParent() : null, ErlangConfigExpression.class);
+        ErlangTupleExpression tuple = ObjectUtils.tryCast(config != null ? config.getParent() : null, ErlangTupleExpression.class);
+        return tuple != null && ObjectUtils.tryCast(tuple.getParent(), oldestParent) != null;
```

Comment:
I know this method and I've read its realization. It takes 2 parameters - class and level. It iterates to necessary level and checks type.
And it will be O(n^2) complexity.